test/controllers/api_users/permissions_controller_test.rb

```diff
@@ -0,0 +1,347 @@
+      application = create(:application, with_supported_permissions: %w[other], with_supported_permissions_not_grantable_from_ui: %w[not_from_ui])
+      api_user = create(:api_user)
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      stub_policy current_user, api_user, update?: true
+
+      not_from_ui_permission = application.supported_permissions.find_by(name: "not_from_ui")
+
+      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [not_from_ui_permission.id] } }
+
+      api_user.reload
+
+      assert_equal [], api_user.supported_permissions
+    end
+
+    should "assign the application id to the application_id flash" do
+      application = create(:application, with_supported_permissions: %w[permission])
+      api_user = create(:api_user)
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      stub_policy current_user, api_user, update?: true
+
+      permission = application.supported_permissions.find_by(name: "permission")
+
+      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [permission.id] } }
+
+      assert_equal application.id, flash[:application_id]
+    end
+
+    should "raise an exception if the user cannot be found" do
+      application = create(:application)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      assert_raises(ActiveRecord::RecordNotFound) do
+        patch :update, params: { api_user_id: "unknown-id", application_id: application, application: { supported_permission_ids: %w[id] } }
+      end
+    end
+
+    should "prevent unauthorised users" do
+      application = create(:application)
+      api_user = create(:api_user)
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      stub_policy current_user, api_user, update?: false
+
+      patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [] } }
+
+      assert_not_authorised
+    end
+
+    should "allow editing permissions for API-only application" do
+      sign_in create(:superadmin_user)
+
+      application = create(:application, api_only: true, with_supported_permissions: %w[permission])
+      api_user = create(:api_user)
+      create(:access_token, resource_owner_id: api_user.id, application:)
+
+      permission = application.supported_permissions.find_by(name: "permission")
+
+      assert_nothing_raised do
+        patch :update, params: { api_user_id: api_user, application_id: application, application: { supported_permission_ids: [permission.id] } }
+      end
+    end
```

Comment:
The name of this test and the setup somewhat implies to me that users can *only* edit API-only applications, but I'm not sure that's the case with the current implementation. If it wsa, then I'd probably expect to see a test demonstrating that editing permissions for a non-API-only app is not allowed.